plaid/src/data/websocket/mod.rs

```diff
@@ -0,0 +1,408 @@
+mod selector;
+
+use crate::executor::Message;
+use crossbeam_channel::Sender;
+use futures_util::{
+    stream::{SplitSink, SplitStream},
+    SinkExt, StreamExt,
+};
+use http::{HeaderMap, Uri};
+use plaid_stl::messages::{Generator, LogSource, LogbacksAllowed};
+use selector::UriSelector;
+use serde::Deserialize;
+use std::{collections::HashMap, str::FromStr, time::Duration};
+use tokio::{net::TcpStream, task::JoinHandle};
+use tokio_tungstenite::{
+    connect_async,
+    tungstenite::{client::IntoClientRequest, protocol::Message as WSMessage},
+    MaybeTlsStream, WebSocketStream,
+};
+
+#[derive(Debug)]
+enum Errors {
+    SocketCreationFailure,
+}
+
+#[derive(Deserialize)]
+pub struct WebsocketDataGenerator {
+    websockets: HashMap<String, Websocket>,
+}
+
+#[derive(Deserialize)]
+pub struct Websocket {
+    /// The URI(s) of the WebSocket endpoint. The config allows multiple URIs in the event that one fails,
+    /// a new one can be used in its place.
+    #[serde(deserialize_with = "parse_uris")]
+    uris: Vec<Uri>,
+    /// A string indicating the type of log associated with the WebSocket.
+    log_type: String,
+    /// The message to be sent over the WebSocket connection.
+    message: Option<String>,
+    /// An optional field containing a map of headers to be included in the WebSocket request.
+    #[serde(deserialize_with = "parse_headers")]
+    headers: Option<HeaderMap>,
+    /// Time (in milliseconds) between sending new messages to the receiving end
+    sleep_duration: u64,
+    /// The number of Logbacks this generator is allowed to trigger
+    #[serde(default)]
+    logbacks_allowed: LogbacksAllowed,
+    /// The minimum amount of time (in milliseconds) to wait before retrying a connection to a WebSocket
+    #[serde(default = "min_retry_duration")]
+    min_retry_duration: u64,
+    /// The maximum amount of time (in milliseconds) to wait before retrying a connection to a WebSocket
+    #[serde(default = "max_retry_duration")]
+    max_retry_duration: u64,
+}
+
+/// The default value for `min_retry_duration` if none is provided in the configuration.
+fn min_retry_duration() -> u64 {
+    100
+}
+
+/// The default value for `max_retry_duration` if none is provided in the configuration.
+fn max_retry_duration() -> u64 {
+    60000
+}
+
+/// Custom parser for URI. Returns an error if an invalid URI is provided
+fn parse_uris<'de, D>(deserializer: D) -> Result<Vec<Uri>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let uris_raw = <Vec<String>>::deserialize(deserializer)?;
+
+    let uris = uris_raw
```

Comment:
Should we also check that the protocol is ws or wss?

Reply:
I think the most we can/should do is throw a warning for sockets whose protocol is ws. After that, it's up to whoever is running the instance to accept that risk